File: cpp/etl/etl.cpp
#include "etl.h"

using namespace std;

namespace etl
{

// Transform record of Scrabble scores from being score index to letter indexed
map<char, int> transform(map<int, vector<char> > legacyScores) {
  map<char, int> reformattedScores;
  for (auto lettersPerScore : legacyScores) {
    int scoreForTheseLetters = lettersPerScore.first;
    vector<char>& letters = lettersPerScore.second;
    for (auto letter : letters) {
      reformattedScores[tolower(letter)] = scoreForTheseLetters;
    }
  }
  return reformattedScores;
}

} // end namespace etl

File: csharp/etl/ETL.cs
using System.Collections.Generic;

public class ETL
{
	// Translate a set of Scrabble scores indexed by point values into a set indexed by letters
	public static Dictionary<string, int> Transform (Dictionary<int, IList<string>> oldScores)
	{
		var newScores = new Dictionary<string, int>();
		// Each old entry has point-value keys with letter values
		foreach (var oldEntry in oldScores) {
			// For each letter with a given score, add it to the transformed output
			foreach(var letter in oldEntry.Value) {
				newScores.Add (letter.ToLower(), oldEntry.Key);
			}
		}
		return newScores;
	}
}
